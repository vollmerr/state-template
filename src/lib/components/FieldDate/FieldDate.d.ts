import * as React from 'react';

export interface FieldDateProps extends HTMLInputElement {
  /** Accessible indicator of related information */
  'aria-describedby'?: string;

  /** Accessible indicator for errors existing */
  'aria-invalid'?: string;

  /** Class names to attach to the field wrapper */
  className?: string;

  /** Disable the input */
  disabled?: boolean;

  /** Id of field */
  id?: string;

  /** Ref to attach to input */
  inputRef?: React.Ref;

  /** Label for the field */
  label: string;

  /** Minimum date able to select */
  minDate?: string | Date;

  /** Name of field */
  name: string;

  /** Called when the date changes */
  onChange?: (string: value?) => void;

  /** Value of the field */
  value?: string;
}

declare class FieldDate extends React.Component<FieldDateProps, {}> {}

export default FieldDate;
