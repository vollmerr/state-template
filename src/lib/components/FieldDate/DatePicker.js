import React, { Component } from 'react';
import T from 'prop-types';
import Pikaday from 'pikaday';

import { isValid } from '../../utils/date';
import Icon from '../Icon';

// Date picker that uses pikaday...
class DatePicker extends Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      displayText: '',
    };
    this.pickerRef = React.createRef();
    this.displayRef = React.createRef();
    this.picker = null;
  }

  componentDidMount() {
    this.initPikaday();
  }

  componentWillReceiveProps(nextProps) {
    const { input } = this.props;
    const newValue = nextProps.input.value;
    if (input.value !== newValue) {
      // update displayed text when value changes
      const displayText = this.getDate(newValue).slice(0, 10);
      this.setState({ displayText });
      // update date in picker
      this.picker.setDate(newValue);
    }
  }

  getDate = (date) => {
    if (isValid(date)) {
      return new Date(date).toISOString();
    }
    return '';
  }

  onSelect = (date) => {
    const { input } = this.props;
    const newDate = this.getDate(date);
    // update redux-form
    input.onChange(newDate);
    // update display field
    this.setState({ displayText: newDate.slice(0, 10) });
  }

  initPikaday = () => {
    const { input, minDate } = this.props;

    this.picker = new Pikaday({
      field: this.pickerRef.current,
      onSelect: this.onSelect,
      minDate,
      yearRange: [new Date().getFullYear(), new Date().getFullYear() + 30],
      trigger: this.displayRef.current,
    });

    this.picker.setDate(input.value);
  }

  render() {
    const {
      input, disabled, minDate, ...rest
    } = this.props;
    const { displayText } = this.state;

    return (
      <div className="field-date has-feedback">
        {/* field that will be displayed */}
        <input
          ref={this.displayRef}
          className={'form-control'}
          disabled={disabled}
          value={displayText}
          onChange={() => {}}
          name={input.name}
          autoComplete={'off'}
        />
        {
          !disabled
          && <Icon name={'calendar'} className={'form-control-feedback'} />
        }
        {/* date picker */}
        <input
          ref={this.pickerRef}
          {...rest}
          className={'hidden'}
        />
      </div>
    );
  }
}

DatePicker.propsTypes = {
  /** redux-form input object */
  input: T.object.isRequired,
  /** picker is disabled */
  disabled: T.bool,
  /** min date to select from */
  minDate: T.oneOfType([T.string, T.instanceOf(Date)]),
};

DatePicker.defaultProps = {
  disabled: false,
  minDate: null,
};

export default DatePicker;
