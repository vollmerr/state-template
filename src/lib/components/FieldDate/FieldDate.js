import React from 'react';
import T from 'prop-types';
import Pikaday from 'pikaday';

import { isInvalidDate } from '../../utils/validate';
import { isValid } from '../../utils/date';
import { withField } from '../Field';
import Icon from '../Icon';

// Date picker for redux-form using pikaday library
export class FieldDate extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      displayText: '',
    };
    this.pickerRef = React.createRef();
    this.displayRef = React.createRef();
    this.picker = null;
  }

  componentDidMount() {
    this.initPikaday();
  }

  componentWillReceiveProps(nextProps) {
    const { value } = this.props;
    if (value !== nextProps.value) {
      // update displayed text when value changes
      const displayText = this.getDate(nextProps.value).slice(0, 10);
      this.setState({ displayText });
      // update date in picker
      this.picker.setDate(nextProps.value);
    }
  }

  getDate = (date) => {
    if (isValid(date)) {
      return new Date(date).toISOString();
    }
    return '';
  }

  onSelect = (date) => {
    const { onChange } = this.props;
    const newDate = this.getDate(date);
    // update passed in value
    if (onChange) {
      onChange(newDate);
    }
    // update display field
    this.setState({ displayText: newDate.slice(0, 10) });
  }

  initPikaday = () => {
    const { value, minDate } = this.props;

    this.picker = new Pikaday({
      minDate,
      field: this.pickerRef.current,
      onSelect: this.onSelect,
      yearRange: [new Date().getFullYear(), new Date().getFullYear() + 30],
      trigger: this.displayRef.current,
    });

    this.picker.setDate(value);
  }

  render() {
    const { disabled, minDate, ...rest } = this.props;
    const { displayText } = this.state;

    return (
      <div data-test={'field--date'} className={'field-date has-feedback'}>
        {/* field that will be displayed */}
        <input
          {...rest}
          disabled={disabled}
          ref={this.displayRef}
          className={'form-control'}
          value={displayText}
          onChange={() => {}}
          onBlur={() => {}}
          onFocus={() => {}}
          autoComplete={'off'}
          data-test={'field--date-control'}
        />

        {
          !disabled
          && (
            <Icon
              data-test={'field--date-icon'}
              name={'calendar'}
              className={'form-control-feedback'}
            />
          )
        }

        {/* date picker */}
        <input
          type={'hidden'}
          ref={this.pickerRef}
          disabled={disabled}
        />
      </div>
    );
  }
}

FieldDate.propTypes = {
  /** Value of date selected */
  value: T.string,

  /** Minimum date able to select */
  minDate: T.oneOfType([T.string, T.instanceOf(Date)]),

  /** Disable the input */
  disabled: T.bool,

  /** Called when date changes */
  onChange: T.func,
};

FieldDate.defaultProps = {
  value: null,
  minDate: null,
  disabled: false,
  onChange: null,
};

export default withField(isInvalidDate)(FieldDate);
