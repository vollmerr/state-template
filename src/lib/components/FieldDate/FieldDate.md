```jsx
const initialValues = {
  disabledDate: '01/02/1969',
};

// example `ref`, access using inputRef.current
// this will be used internally instead of the displayRef
const inputRef = React.createRef();

<ExampleForm form={'date'} initialValues={initialValues} inputRef={inputRef}>
  <FieldDate
    name={'date'}
    label={'Date'}
    helpText={'This is the `helpText` content...'}
    tooltip={'Example tooltip!'}
    inputRef={inputRef}
  />
  <FieldDate 
    required
    name={'requiredDate'}
    label={'Required Date'}
  />
  <FieldDate
    name={'validationsDate'}
    label={'Date with Custom Validations'}
    validate={[
      (val) => new Date(val) >= new Date()
        ? 'Value must be in the past' 
        : undefined
    ]}
  />
  <FieldDate
    minDate={new Date()}
    name={'minDate'}
    label={'Date with Min Date'}
    helpText={'Min date of today'}
  />
  <FieldDate
    disabled
    name={'disabledDate'}
    label={'Disabled Date'}
  />
</ExampleForm>
```
