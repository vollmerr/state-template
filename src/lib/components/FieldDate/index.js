export { default as DatePicker } from './DatePicker';
export { default } from './FieldDate';
