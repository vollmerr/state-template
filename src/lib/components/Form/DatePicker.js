import React, { Component } from 'react';
import Pikaday from 'pikaday';

class DatePicker extends Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      displayText: '',
    };
    this.pickerRef = React.createRef();
    this.displayRef = React.createRef();
    this.picker = null;
  }

  componentDidMount() {
    this.initPikaday();
  }

  componentWillReceiveProps(nextProps) {
    // handle form resets
    if (!nextProps.input.value) {
      this.setState({ displayText: '' });
    }
  }

  onSelect = (date) => {
    const { input } = this.props;
    const newDate = new Date(date).toISOString();
    // update redux-form
    input.onChange(newDate);
    // update display field
    this.setState({ displayText: newDate.slice(0, 10) });
  }

  initPikaday = () => {
    const { input, minDate } = this.props;

    this.picker = new Pikaday({
      field: this.pickerRef.current,
      onSelect: this.onSelect,
      minDate,
      yearRange: [new Date().getFullYear(), new Date().getFullYear() + 30],
      trigger: this.displayRef.current,
    });

    if (input.value) {
      this.picker.setDate(input.value);
    }
  }

  render() {
    const {
      input, minDate, disabled, ...props
    } = this.props;
    const { displayText } = this.state;

    return (
      <div className="fieldDate has-feedback">
        {/* field that will be displayed */}
        <input
          ref={this.displayRef}
          className="form-control"
          disabled={disabled}
          value={displayText}
          onChange={() => {}}
        />
        {!disabled && <span className="ca-gov-icon-calendar form-control-feedback" />}
        {/* date picker */}
        <input
          ref={this.pickerRef}
          {...props}
          className="hidden"
        />
        {/* handle dispatching to redux-form */}
        <input
          className="hidden"
          id={input.name}
          onChange={(v) => { console.log('changed: ', v); }}
        />
      </div>
    );
  }
}

export default DatePicker;
