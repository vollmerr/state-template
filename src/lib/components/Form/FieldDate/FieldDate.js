import React from 'react';
import T from 'prop-types';

import { isInvalidDate } from '../../../utils/validate';

import FieldWrapper from '../FieldWrapper';
import withField from '../withField';
import DatePicker from './DatePicker';

// Date picker for redux-form using pikaday library
export const FieldDate = (props) => {
  const {
    input, disabled, minDate, ...rest
  } = props;

  return (
    <FieldWrapper data-test={'field-date'} disabled={disabled} {...rest}>
      <DatePicker input={input} disabled={disabled} minDate={minDate} />
    </FieldWrapper>
  );
};

FieldDate.propTypes = {
  /** Input from redux-form's Field, attaches name, value, etc  */
  input: T.object.isRequired,
  /** Disable the input */
  disabled: T.bool,
  /** Minimum date able to select */
  minDate: T.oneOfType([T.string, T.instanceOf(Date)]),
};

FieldDate.defaultProps = {
  disabled: false,
  minDate: null,
};

const withReduxField = withField(isInvalidDate);
export default withReduxField(FieldDate);
