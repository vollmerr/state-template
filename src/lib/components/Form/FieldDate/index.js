import DatePicker from './DatePicker';
import FieldDate from './FieldDate';

export {
  DatePicker,
};

export default FieldDate;
