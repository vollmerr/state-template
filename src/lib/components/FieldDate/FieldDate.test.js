import React from 'react';
import { shallow } from 'enzyme';

import DatePicker from './DatePicker';
import { FieldDate } from './FieldDate';

const { testUtils } = global;

const props = {
  input: testUtils.getInputProp(),
  meta: testUtils.getMetaProp(),
  label: 'test label',
  minDate: '1900-01-02',
  disabled: false,
};

let wrapper;
describe('FieldDate', () => {
  beforeEach(() => {
    wrapper = shallow(<FieldDate {...props} />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a `DatePicker`', () => {
    expect(wrapper.find(DatePicker).length).toBe(1);
  });
});
