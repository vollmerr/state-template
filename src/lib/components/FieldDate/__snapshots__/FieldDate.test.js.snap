// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldDate should render correctly 1`] = `
<div
  className="field__date test-classname"
  data-test="field__date"
>
  <input
    aria-describedby="aria-desc-id"
    aria-invalid="false"
    autoComplete="off"
    className="form-control"
    data-test="field__date-control"
    disabled={false}
    id="test-id"
    label="test-label"
    name="test-name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value="2020-05-02"
  />
  <Icon
    className="field__date-icon"
    data-test="field__date-icon"
    name="calendar"
    tag="span"
  />
  <input
    disabled={false}
    type="hidden"
  />
</div>
`;
